Pass a program object to the linter to enable type checking

smockle opened the pull request:
Attempting to fix #109 (building on @zaggino’s code snippet in that issue)

# Introduction
This PR enables [semantic rules](https://palantir.github.io/tslint/usage/type-checking/) by passing a program object to the linter to enable type checking. This resolves warnings that appear in the developer tools console (<kbd>Command</kbd><kbd>Option</kbd><kbd>i</kbd>) when certain TSLint rules are used, e.g.:
```
Warning: The 'no-unused-variable' rule requires type checking
```

# Warning
This code makes `linter-tslint` runs very slowly on my machine. Warnings like this appear:
```
Handling of 'mousewheel' input event was delayed for 16200 ms due to main thread being busy. Consider marking event handler as 'passive' to make the page more responive.
```
I’m not sure how to resolve this. Any suggestions are welcomed.

Xapphire13 commented on lib/main.js:
```diff
@@ -13,6 +13,17 @@ let tslintDef;
 let requireResolve;
 const idleCallbacks = new Set();
 
+async function statAsync(pathname) {
```
Maybe name this just `stat` to keep inline with the [api](https://nodejs.org/api/fs.html#fs_fs_stat_path_callback)

smockle replied:
Fixed in https://github.com/AtomLinter/linter-tslint/pull/157/commits/94ad4d1871edd5b6b3174c8af1d3e9c675ec6e56

Xapphire13 commented on lib/main.js:
```diff
@@ -235,10 +261,19 @@ export default {
-    }, options));
+    const program = await this.getProgram(Linter, configurationPath);
+    let linter;
+    if (program) {
```
If the `if/else` needed? Since passing null/undefined is effectively the same as not passing at all

smockle replied:
Nope! Thanks. 😊 Removed in https://github.com/AtomLinter/linter-tslint/pull/157/commits/60be059ed7ad45ed87eb78ed6ac1e0eb94c252e9

Xapphire13 commented on lib/main.js:
```diff
@@ -189,6 +200,21 @@ export default {
 
+  async getProgram(Linter, configurationPath) {
+    let program;
+    const tsconfigPath = path.resolve(path.dirname(configurationPath), 'tsconfig.json');
```
Instead of trying to check if `tsconfig.json` exists under `configurationPath`, I would use [`Linter.findConfigurationPath(null, configurationPath)`](https://github.com/palantir/tslint/blob/master/src/configuration.ts#L116) to fetch the config file, this way it will call TSLint's logic for searching up the hierarchy; and we get the search code for free (can remove the [`statAsync()`](https://github.com/AtomLinter/linter-tslint/pull/157/files#diff-16b7d75b70953cbf4755508170f2f24cR16) method you added)

smockle replied:
It looks like`Linter.findConfigurationPath` searches for a `tslint.json` (i.e. `CONFIG_FILENAME`) rather than a project’s `tsconfig.json`, which is what TSLint needs to create a program object: https://github.com/palantir/tslint/blob/617e705330a4e64f84d8d2759f684437203b75a8/src/linter.ts#L56-L59

Is there sample code demonstrating how to locate `tsconfig.json` using `Linter.findConfigurationPath`?

Xapphire13 replied:
Oops, my bad. My brain totally read `tsconfig.json` as `tslint.json` haha. Disregard my suggestion =]. I will create a task to eventually mimick the search up logic for the `tsconfig.json` file at a later date.

Xapphire13 commented:
I wonder if the slowness issue would be fixed if the linting was done asynchronously. I'm planning on updating the `linter-tslint` code to perform the linking in a task process soon, we can check the performance of this when that refactor is complete.

Xapphire13 commented on lib/main.js:
```diff
@@ -189,6 +200,21 @@ export default {
+    } catch (err) {
+      // no-op
+    }
+    return Promise.resolve(program);
```
No need to wrap in a promise as the async function will automatically return a promise

smockle replied:
Fixed in https://github.com/AtomLinter/linter-tslint/pull/157/commits/9bed459f8572a3213893836b2a8e55dd700a0d1b

Arcanemagus commented on lib/main.js:
```diff
@@ -235,10 +261,11 @@ export default {
       }
     }
 
+    const program = await this.getProgram(Linter, configurationPath);
```
This probably needs to be in a wrapper checking whether `configurationPath` is a valid value, depending on what `Linter.findConfigurationPath` returns when it _can't_ find one. You are using this as an argument to `path.dirname` above which will throw a deprecation error if it isn't a valid string.

smockle replied:
Fixed in https://github.com/AtomLinter/linter-tslint/pull/157/commits/5de77d83408e5d656c5e994729764daafbe38304

Arcanemagus commented on lib/main.js:
```diff
@@ -189,6 +200,21 @@ export default {
+      const stats = await stat(tsconfigPath);
+      const isFile = stats.isFile();
+      if (isFile) {
+        program = Linter.createProgram('tsconfig.json', path.dirname(configurationPath));
```
You are determining `path.dirname(configurationPath)` twice, just store it in a `const`.

smockle replied:
Fixed in https://github.com/AtomLinter/linter-tslint/pull/157/commits/8c8f7f70e20332d44bb5f72bafe27e635298dd5b

Arcanemagus commented:
Can we get a spec testing that this is working properly?

Xapphire13 commented:
Also, could you add a setting so that this is an opt in feature. That way people aren't impacted by the bad perf, unless they configure it to be on in the settings. We can remove this if the perf gets better in the future (with worker tasks etc)

Arcanemagus commented:
Wait, TSLint does the delay even when there is no program to compile?

smockle commented:
@Xapphire13 Opt-in setting added in https://github.com/AtomLinter/linter-tslint/pull/157/commits/5e7bb979e35007bd3bdf1a18e907ba07cac2aaa3

@Arcanemagus Specs added in https://github.com/AtomLinter/linter-tslint/pull/157/commits/b526e21658703971d6ef98ecdea5685380aa4a0e and modified in https://github.com/AtomLinter/linter-tslint/pull/157/commits/5e7bb979e35007bd3bdf1a18e907ba07cac2aaa3 to support the new opt-in setting

smockle commented:
> Wait, TSLint does the delay even when there is no program to compile?

I don’t think so. What’s the best way to time this?

I interpreted @Xapphire13’s request for an opt-in setting to handle the case where a `tsconfig.json` does exist—usually the presence of a `tsconfig.json` triggers code to create and pass a program, but this will be skipped if the `enableSemanticRules` is `false`.

I‘m not sure which part is slow—it might be checking for the `tsconfig.json`, creating the program (`Linter.createProgram`) or extra processing the linter does when it’s passed a program. I suspect the latter.

Xapphire13 commented:
Yes, my request was to bypass the program logic when there _is_ a tsconfig.json file. This way no one is impacted (without choosing to be) until we get the perf sorted.

smockle commented:
I just pushed a branch with analogous changes, based on PR #158: https://github.com/smockle/linter-tslint/tree/feature/type-checking-workerTask

Xapphire13 commented:
How's the perf?

Xapphire13 commented on spec/fixtures/invalid-typechecked/tslint.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "rules": {
+    "no-unused-variable": true
```
[This rule](https://palantir.github.io/tslint/rules/no-unused-variable/) doesn't require type-checking.

smockle replied:
Thanks! Is that still true as of TSLint 5.x (which `linter-tslint` [depends on](https://github.com/AtomLinter/linter-tslint/blob/master/package.json#L45))? From the [changelog](https://github.com/palantir/tslint/blob/d8db2467b745513d3a3d02b197f46009cd685d81/CHANGELOG.md#api-breaks-for-custom-rules):
> This means that some rules that previously worked without the type checker now require it. This includes:
> * no-unused-variable

smockle replied:
Related, [TSLint’s tests](https://github.com/palantir/tslint/tree/master/test/rules/no-unused-variable) for `no-unused-variable` use the `tslint.json` below. Should I add `"linterOptions": { "typeCheck": true }` to the `tslint.json` files I added in this PR?

```JSON
{
  "linterOptions": {
    "typeCheck": true
  },
  "rules": {
    "no-unused-variable": true
  }
}
```

Xapphire13 replied:
I was thinking of adding a rule that the current documentation lists as needing type-checking. But I'm not strongly opposed to leaving it as is if you wish.

smockle replied:
Oh, I hadn’t even considered that. That does make a lot of sense. I’ll update the PR. Thanks!

Xapphire13 replied:
Cool =]. Bear in mind that I checked in the worker task code. So you'll need to merge that change in.

smockle replied:
Fixed in https://github.com/AtomLinter/linter-tslint/pull/157/commits/f663517f416cd0f9c68d9c3c54c6e113be83b783

smockle commented:
> How's the perf?

It seemed better! I still get messages like
```
Handling of 'mousewheel' input event was delayed for <n> ms due to main thread being busy. Consider marking event handler as 'passive' to make the page more responive.
```
but `<n>` is around `~300ms` rather than `~16000ms`. Orders of magnitude of improvement. Great work! 😊 I realize this is anecdotal, but I don’t know how to reliably profile Atom extension performance.

smockle commented:
Rebased to lastest master so this merges cleanly with the worker task code. Updated the spec to test [`no-boolean-literal-compare`](https://palantir.github.io/tslint/rules/no-boolean-literal-compare/) rather than `no-unused-variable`.

Arcanemagus commented on lib/worker.js:
```diff
@@ -94,6 +106,22 @@ async function getLinter(filePath) {
+  const tsconfigPath = path.resolve(configurationDir, 'tsconfig.json');
+  try {
+    const stats = await stat(tsconfigPath);
+    const isFile = stats.isFile();
```
Since this isn't re-used, and the function name is already clear enough you might as well just move this directly into the conditional.

Arcanemagus commented on package.json:
```diff
@@ -26,6 +26,12 @@
+    "enableSemanticRules": {
+      "type": "boolean",
+      "title": "Enable semantic rules",
+      "description": "Pass a TypeScript program object to the linter. May negatively affect performance. See this page for details: https://palantir.github.io/tslint/usage/type-checking/",
```
"Allow passing a..."

This should be conditionally worded since it doesn't forcibly pass one every time.

smockle commented:
Awesome—thank you for reviewing and for addressing those nits!
